sycl/include/sycl/vector.hpp

```diff
@@ -144,17 +156,34 @@ template <typename DataT, int NumElements> class vec_base {
   static constexpr size_t AdjustedNum = (NumElements == 3) ? 4 : NumElements;
   // This represent type of underlying value. There should be only one field
   // in the class, so vec<float, 16> should be equal to float16 in memory.
-#if defined(__INTEL_PREVIEW_BREAKING_CHANGES) &&                               \
-    defined(__SYCL_USE_NEW_VEC_IMPL)
-  using DataType = DataT[AdjustedNum];
+  //
+  // In intel/llvm#14130 we incorrectly used std::array as an underlying storage
+  // for vec data. The problem with std::array is that it comes from the C++
+  // STL headers which we do not control and they may use something that is
+  // illegal in SYCL device code. One of specific examples is use of debug
+  // assertions in MSVC's STL implementation.
+  //
+  // The better approach is to use plain C++ array, but the problem here is that
+  // C++ specification does not provide any guarantees about the memory layout
+  // of std::array and therefore directly switching to it would technically be
+  // an ABI-break, even though the practical chances of encountering the issue
+  // are low.
+  //
+  // To play it safe, we only switch to use plain array if both its size and
+  // alignment match those of std::array, or unless the new behavior is forced
+  // via __SYCL_USE_NEW_VEC_IMPL or preview breaking changes mode.
+  using DataType = std::conditional_t<
+#if __SYCL_USE_NEW_VEC_IMPL
```

Comment:
FYI, I think @aelovikov-intel is also using this flag for some other (draft?) `sycl::vec` refactoring work. So, if the user explicitly specifies `__SYCL_USE_NEW_VEC_IMPL` macro, it might bring in Andrei's changes as well, along with the switch to using C-arrays.

Reply:
The main point here is to have some sort of opt-in mechanism for customers who value building in debug mode on windows using `clang.exe` (instead of `clang-cl.exe`) over backwards compatibility or potential ABI issues. So, I'm fine with choosing any other name here, I've just re-used what we had already

Reply:
Done in 0c1c53d61ae8ad19f6ba50c7de2256609eebb48e